dynamic_programming/levenshtein_distance.nim

```diff
@@ -0,0 +1,78 @@
+## Levenshtein distance
+
+import tables
+
+type Key = (Natural, Natural)
+
+
+func toKey(indA, indB: int): Key =
+  return (indA.Natural, indB.Natural)
+
+
+func initSubsolutions(lenA, lenB: Natural): Table[Key, Natural] =
+  for indA in 0.Natural..lenA:
+    result[toKey(indA, 0)] = indA
+
+  for indB in 1.Natural..lenB:
+    result[toKey(0, indB)] = indB
+
+
+func computeLevenshteinDistanceMatrix(a, b: string): Table[Key, Natural] =
+  result = initSubsolutions(a.len, b.len)
```

Comment:
> I decided to leave the initSubsolutions function (in order to make computeLevenshteinDistanceMatrix shorter).

The issue with `initSubsolutions` is that it doesn't really init the table, the object is left in partially filled state and is unusable by itself as returned from the function. This suggests that it either has to be an proc-local function, just a block, or, rather, just 4 more lines in `computeLevenshteinDistanceMatrix` that you then can justifiably rename to `initLevenshteinDistanceMatrix`.

```suggestion
func computeLevenshteinDistanceMatrix(a, b: string): Table[Key, Natural] =
  # Partially init the matrix:
  # - Pre-fill with distances between all prefixes of `a` and an empty string
  for indA in 0.Natural..a.len:
    result[toKey(indA, 0)] = indA
  # - Pre-fill with distances between an empty string and all prefixes of `b`
  for indB in 1.Natural..b.len:
    result[toKey(0, indB)] = indB
```

Reply:
Please have a look at 9aa9e251d3cb3d08019f66c45677c383ba151c68 and 2ccfb918577a6173e14897a9529db650d9b15eab.